### PR TITLE
(PUP-10295) Add range intersection support to gem package provider

### PR DESCRIPTION
**Pull request by luchihoratiu**
Add support for range intersection (min,max such as: `>=A <=B`) specified in the :ensure manifest field. This will install a version that satisfies the provided range.

### CODE REVIEW

**Comment by luchihoratiu** [comment]
jenkins please test this on centos7-64a,redhat7-64a

File: lib/puppet/provider/package/gem.rb [hunk 1]
```diff
@@ -17,6 +19,9 @@
 
   has_feature :versionable, :install_options, :uninstall_options, :targetable, :version_ranges
 
+  GEM_VERSION       = Puppet::Util::Package::Version::Gem
+  GEM_VERSION_RANGE = Puppet::Util::Package::Version::Range
+
```
Comment (luchihoratiu) [inline lib/puppet/provider/package/gem.rb]: These cause warnings for some reason:
> /Users/luchi/repos/puppet/lib/puppet/provider/package/gem.rb:22: warning: already initialized constant GEM_VERSION
/Users/luchi/repos/puppet/lib/puppet/provider/package/gem.rb:22: warning: previous definition of GEM_VERSION was here
/Users/luchi/repos/puppet/lib/puppet/provider/package/gem.rb:23: warning: already initialized constant GEM_VERSION_RANGE
/Users/luchi/repos/puppet/lib/puppet/provider/package/gem.rb:23: warning: previous definition of GEM_VERSION_RANGE was here

Tried with different names but that doesn't help. It acts like the file is being loaded twice. 
`pry-byebug` got me to `puppet/lib/puppet/util/autoload.rb:80` but I couldn't figure the root cause yet. @puppetlabs/night-s-watch ideas are very welcomed 😅

**Comment by luchihoratiu** [comment]
jenkins please test this on centos7-64a,redhat7-64a

**Comment by luchihoratiu** [comment]
jenkins please test this on centos7-64a,redhat7-64a

**Comment by luchihoratiu** [comment]
jenkins please test this on centos7-64a,redhat7-64a

File: lib/puppet/provider/package/gem.rb (hunk 1 above)
Comment (luchihoratiu) [reply lib/puppet/provider/package/gem.rb]: After discussing about this, it was decided to remove these variables to avoid the warnings and track the found issue in [PUP-10390](https://tickets.puppetlabs.com/browse/PUP-10390).

File: lib/puppet/provider/package/gem.rb [hunk 2]
```diff
@@ -125,9 +127,30 @@ def self.gemsplit(desc)
 
   def insync?(is)
     return false unless is && is != :absent
+    should = @resource[:ensure]
 
+    unless should =~ /,/ or should =~ Gem::Requirement::PATTERN
```
Comment (gimmyxd) [inline lib/puppet/provider/package/gem.rb]: In ruby, in conditions, `||` is used over  `or` as they behave differently

https://zverok.github.io/blog/2016-02-23-control-flow.html

File: lib/puppet/provider/package/gem.rb [hunk 3]
```diff
@@ -125,9 +127,30 @@ def self.gemsplit(desc)
...
+      rescue Puppet::Util::Package::Version::Range::ValidationFailure, Puppet::Util::Package::Version::Gem::ValidationFailure
+        Puppet.debug("Cannot parse #{should} as a ruby gem version range")
+        return false
+      end
+
+      begin
+        is.each do |is_installed|
+          is_version = Puppet::Util::Package::Version::Gem.parse(is_installed)
+          should_install = should_range.include?(is_version)
+          return true if should_install
```
Comment (gimmyxd) [inline lib/puppet/provider/package/gem.rb]:
```ruby
return true if should_range.include?(is_version)
```

File: lib/puppet/provider/package/gem.rb [hunk 4]
```diff
@@ -148,12 +171,24 @@ def install(useversion = true)
     command = resource_or_provider_command
     command_options = ["install"]
     command_options += install_options if resource[:install_options]
+    should = resource[:ensure]
+
+    unless should =~ /,/ or should =~ Gem::Requirement::PATTERN
```
Comment (gimmyxd) [inline lib/puppet/provider/package/gem.rb]: same, use `||` instead of `or`

File: lib/puppet/provider/package/gem.rb [hunk 5]
```diff
@@ -148,12 +171,24 @@ def install(useversion = true)
...
+      begin
+        should_range = Puppet::Util::Package::Version::Range.parse(should, Puppet::Util::Package::Version::Gem)
+        if should_range.is_a?(Puppet::Util::Package::Version::Range::MinMax)
+          should = should_range.to_s.sub(' ', ',')
+          useversion = true
```
Comment (gimmyxd) [inline lib/puppet/provider/package/gem.rb]: why is this needed only on `MinMax`

File: lib/puppet/provider/package/gem.rb [hunk 6]
```diff
@@ -148,12 +171,24 @@ def install(useversion = true)
...
+        end
+      rescue Puppet::Util::Package::Version::Range::ValidationFailure, Puppet::Util::Package::Version::Gem::ValidationFailure
+        Puppet.debug("Cannot parse #{should} as a ruby gem version range. Falling through.")
+      end
+    end
 
     if Puppet::Util::Platform.windows?
-      version = resource[:ensure]
-      command_options << "-v" << %Q["#{version}"] if (! resource[:ensure].is_a? Symbol) and useversion
+      command_options << "-v" << %Q["#{should}"] if useversion and !should.is_a? Symbol
```
Comment (gimmyxd) [inline lib/puppet/provider/package/gem.rb]: you can also replace `and` with `&&`

File: lib/puppet/provider/package/gem.rb [hunk 7]
```diff
@@ -148,12 +171,24 @@ def install(useversion = true)
...
     else
-      command_options << "-v" << resource[:ensure] if (! resource[:ensure].is_a? Symbol) and useversion
+      command_options << "-v" << should if useversion and !should.is_a? Symbol
```
Comment (gimmyxd) [inline lib/puppet/provider/package/gem.rb]: you can also replace `and` with `&&`

File: lib/puppet/util/package/version/gem.rb [hunk 8]
```diff
@@ -0,0 +1,15 @@
+module Puppet::Util::Package::Version
+  class Gem < ::Gem::Version
+    def self.parse(version)
+      raise ValidationFailure, version.to_s unless version.is_a? String
```
Comment (gimmyxd) [inline lib/puppet/util/package/version/gem.rb]: no need to use `to_s` as you interpolate in the` initialize` of `ValidationFailure`

**Comment by luchihoratiu** [comment]
jenkins please test this on centos7-64a,redhat7-64a

File: acceptance/tests/provider/package/gem.rb [hunk 9]
```diff
@@ -79,6 +80,34 @@
...
+
+      # Install besides existing package (with version between 0.7 and 0.8.1) and expect changes
+      apply_manifest_on(agent, package_manifest2, :expect_changes => true) do
+        list = on(agent, "#{puppet_gem_command} list").stdout
+        match = list.match(/#{package} \((.+)\)/)
+        installed_versions = match[1] if match
+        assert_match(installed_versions, '0.8.0, 0.6.0')
+      end
+      on(agent, "#{puppet_gem_command} uninstall #{package} --all")
+    end
```
Comment (ciprianbadescu) [inline acceptance/tests/provider/package/gem.rb]: the test should leave the system in the same state it was found => package should be uninstalled if it was not present at the beginning

File: acceptance/tests/provider/package/gem.rb [hunk 10]
```diff
@@ -79,6 +80,34 @@
...
+
+      # Ensure the package is not already installed
+      on(agent, "#{puppet_gem_command} uninstall #{package}", :accept_all_exit_codes => true)
+
+      # Make a fresh package install (with version between 0.5 and 0.7)
+      apply_manifest_on(agent, package_manifest1, :expect_changes => true) do
+        list = on(agent, "#{puppet_gem_command} list").stdout
+        match = list.match(/#{package} \((.+)\)/)
+        installed_version = match[1] if match
+        assert_match(installed_version, '0.6.0')
```
Comment (ciprianbadescu) [inline acceptance/tests/provider/package/gem.rb]: could we have only one assert, something like (not tested): `assert_match(list, "#{package} (0.6.0)")`

File: lib/puppet/provider/package/gem.rb [hunk 11]
```diff
@@ -125,9 +127,29 @@ def self.gemsplit(desc)
 
   def insync?(is)
     return false unless is && is != :absent
+    should = @resource[:ensure]
 
+    unless should =~ /,/ || should =~ Gem::Requirement::PATTERN
```
[the remark this answers is not in the document]
Comment (ciprianbadescu) [reply lib/puppet/provider/package/gem.rb]: or in one regexp `should =~ /(?:,|Gem::Requirement::PATTERN)/

File: lib/puppet/provider/package/gem.rb [hunk 12]
```diff
@@ -148,12 +170,22 @@ def install(useversion = true)
     command = resource_or_provider_command
     command_options = ["install"]
     command_options += install_options if resource[:install_options]
+    should = resource[:ensure]
+
+    unless should =~ /,/ || should =~ Gem::Requirement::PATTERN
```
Comment (ciprianbadescu) [inline lib/puppet/provider/package/gem.rb]: in one regexp `should =~ /(?:,|Gem::Requirement::PATTERN)/`

File: lib/puppet/provider/package/gem.rb [hunk 13]
```diff
@@ -148,12 +170,22 @@ def install(useversion = true)
...
+      begin
+        should_range = Puppet::Util::Package::Version::Range.parse(should, Puppet::Util::Package::Version::Gem)
+        should = should_range.to_s.sub(' ', ',')
```
Comment (ciprianbadescu) [inline lib/puppet/provider/package/gem.rb]: I assume here is the transformation from Version::Range to Gem, For example `>1 <2` to `>1,<2`.
If this it is, it should be part of a Version::Range method, `to_gem_spec` or something, as we should not make here assumptions about `Version::Range` internals.

**Comment by ciprianbadescu** [comment]
LGTM, made some smoke tests and they passed. OK to merge after review comments are solved

**Comment by luchihoratiu** [comment]
jenkins please test this on centos7-64a,redhat7-64a
